### PR TITLE
Allow user to delete self

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

- Provide the means for a user to delete own account under the following rules:
   - user is not the last platform administrator
   - user is not the sole owner of a team
- Adds DELETE route to `user`
- Adds deletion button to "User Settings"
- Adds logger for user.deleted
- Updates UI for logging
- Updates failing tests (due to above rules changes)
- Adds the following tests...
   - Can not delete own account if cookie not present
   - Admin can delete own account if another admin exists
   - Last admin cannot delete own account
   - Last owner of a team cannot delete own account
   - Non admin user who is a team member can delete own account
   - Team owner can delete own account if another team owner exists
- Updates "User Management" doc

### Deleting last admin
image

image

### Deleting sole team owner
image

image

## Related Issue(s)

#1767

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass
 - [x] Documentation has been updated
    - [-] Upgrade instructions
    - [-] Configuration details
    - [x] Concepts

## Labels

 - [-] Backport needed? -> add the `backport` label
 - [-] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: forge/db/models/User.js [hunk 1]
```diff
@@ -236,7 +251,7 @@ module.exports = {
                     return M.TeamMember.getTeamMembership(this.id, teamId, includeTeam)
                 },
                 getTeamsOwned: async function () {
-                    return M.TeamMember.getTeamsOwnedBy(this.id)
+                    return await M.TeamMember.getTeamsOwnedBy(this.id)
```
Comment (Steve-Mcl) [inline forge/db/models/User.js:254]:
```suggestion
                    return M.TeamMember.getTeamsOwnedBy(this.id)
```
revert temporary debugging change

File: docs/admin/user_management.md [hunk 2]
```diff
@@ -20,4 +20,10 @@ Users can only be removed if they don't own any teams and projects. As such the
 user should delete any projects and ensure their teams have alternative owners. They
 can either do this themselves or an Admin user can do it for them.
 
-Once all teams and projects are removed, the user can be removed using the option in the "Edit User" dialog.
+### Deleting your own account
```
Comment (knolleary) [inline docs/admin/user_management.md:23]: This section is for admin users - not where regular users will go looking for help.

We probably want to add a bit to `docs/cloud/README.md` as well.

File: forge/routes/api/user.js [hunk 3]
```diff
@@ -97,4 +97,36 @@ module.exports = async function (app) {
...
+            // is already deleted at this point
+            await app.auditLog.User.user.deleted(0, null, deletedUser)
+            reply.clearCookie('sid')
+            reply.send({ status: 'okay' })
+        } catch (err) {
+            const resp = { code: 'unexpected_error', error: err.toString() }
+            await app.auditLog.User.user.deleted(request.session.User, resp, request.session.User)
+            reply.code(400).send(resp)
+        }
+        return reply // avoid "Promise may not be fulfilled with 'undefined' when statusCode is not 204" https://github.com/fastify/help/issues/627
```
Comment (knolleary) [inline forge/routes/api/user.js:130]: Trying to figure out why this is needed, given we don't have to do this in any other route.

File: frontend/src/components/audit-log/AuditEntryVerbose.vue [hunk 4]
```diff
@@ -174,7 +174,12 @@
         <span v-if="!error && entry.trigger.user && entry.body.user">User '{{ entry.trigger.user?.name }}' has created a new user, '{{ entry.body.user?.name }}'.</span>
         <span v-else-if="!error">User data not found in audit entry.</span>
     </template>
-    <template v-else-if="entry.event === 'user.deleted-user'">
+    <template v-else-if="entry.event === 'user.deleted'">
+        <label>{{ AuditEvents[entry.event] }}</label>
+        <span v-if="!error && entry.body?.user">User '{{ entry.body?.user?.name }}' has deleted deleted their account.</span>
```
Comment (knolleary) [inline frontend/src/components/audit-log/AuditEntryVerbose.vue:179]:
```suggestion
        <span v-if="!error && entry.body?.user">User '{{ entry.body?.user?.name }}' has deleted their account.</span>
```

File: frontend/src/pages/account/Settings.vue [hunk 5]
```diff
@@ -19,7 +19,10 @@
             </div>
         </template>
         <template v-else>
-            <ff-button @click="startEdit">Edit</ff-button>
+            <div class="flex space-x-4">
```
Comment (knolleary) [inline frontend/src/pages/account/Settings.vue:22]: The way this is presented, it feels like a 'cancel' button next to 'edit'.

I think it should appear more consistent with the 'Delete Project' button under Project/Instance settings:

image

The text can also say something about needing to delete/transfer team ownership as well. We shouldn't let the user try to delete their account if we know it's going to fail.

File: forge/routes/api/user.js (hunk 3 above)
Comment (Steve-Mcl) [reply forge/routes/api/user.js:130]: Trying to cast my mind back to when I encountered the `"Promise may not be fulfilled"` problem.  I will be honest and state I did not try this route without the `return reply` (I added it knowing it wont break anything and _may_ solve a _possible_ issue.)

I will it and see what the tests comeback with.

File: frontend/src/pages/account/Settings.vue (hunk 5 above)
Comment (Steve-Mcl) [reply frontend/src/pages/account/Settings.vue]: updated in [7c26d2c](https://github.com/flowforge/flowforge/pull/1787/commits/7c26d2c2384c071a5586e301c08e35043bac3e07)

image

File: forge/routes/api/user.js (hunk 3 above)
Comment (Steve-Mcl) [reply forge/routes/api/user.js]: removed in [0503989](https://github.com/flowforge/flowforge/pull/1787/commits/05039891a2b7fb1b6a2597468cd518389284a85f)

File: docs/admin/user_management.md (hunk 2 above)
Comment (Steve-Mcl) [reply docs/admin/user_management.md]: updated in [b503d52](https://github.com/flowforge/flowforge/pull/1787/commits/b503d521e8a6c2bd9b5fc435c6605106b7a06377)

**Comment by Steve-Mcl** [comment]
tests failing on untouched / unrelated issue: 
```
/home/runner/work/flowforge/flowforge/frontend/src/pages/project/Overview.vue
Error:   107:9  error  The "PlusSmIcon" component has been registered but not used  vue/no-unused-components
```

**Comment by Pezmc** [comment]
@Steve-Mcl Fix in https://github.com/flowforge/flowforge/pull/1814

**Comment by Steve-Mcl** [comment]
Thanks @Pezmc   weird that even though #1814 is merged, it still fails. 

Will merge or rebase shortly.

**Review by hardillb — APPROVED** [review]
Looks good to me

**Review by knolleary — APPROVED** [review]

**Pull request merged**